From 2599bbda11b42cc9e38043b70d9e7af2be89e62e Mon Sep 17 00:00:00 2001
From: Lubomir Rintel <lkundrak@v3.sk>
Date: Fri, 6 Jun 2014 18:03:03 +0200
Subject: [PATCH 4/8] Add possibility to set up a loopback device with size and
 offset

Useful for partitions. Also, we can't touch the underlying storage then,
treating the loopback volume as a physical disk partition.
---
 imgcreate/fs.py | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/imgcreate/fs.py b/imgcreate/fs.py
index 378b49b..98cc9e1 100644
--- a/imgcreate/fs.py
+++ b/imgcreate/fs.py
@@ -283,12 +283,17 @@ class RawDisk(Disk):
 
 class LoopbackDisk(Disk):
     """A Disk backed by a file via the loop module."""
-    def __init__(self, lofile, size):
+    def __init__(self, lofile, size, offset=0):
         Disk.__init__(self, size)
         self.lofile = lofile
+        self.size = size
+        self.offset = offset
 
     def fixed(self):
-        return False
+        if self.offset:
+            return True
+        else:
+            return False
 
     def exists(self):
         return os.path.exists(self.lofile)
@@ -308,7 +313,8 @@ class LoopbackDisk(Disk):
         device = losetupOutput.split()[0]
 
         logging.info("Losetup add %s mapping to %s"  % (device, self.lofile))
-        rc = call(["/sbin/losetup", device, self.lofile])
+        rc = call(["/sbin/losetup", "-o", str(self.offset),
+                   "--sizelimit", str(self.size), device, self.lofile])
         if rc != 0:
             raise MountError("Failed to allocate loop device for '%s'" %
                              self.lofile)
-- 
1.9.3

